### DISCUSSION TITLE
Thing registration, looking for best-practices

### DISCUSSION THREAD

**Discussion by lathoub** [FraunhoferIOSB/FROST-Server#843] [Q&A]
### Context/situation
For a `Thing` to send `Observations` to the service, it needs the `Datastream @iot.id` . This identifier should not be baked into he Thing's firmware, it should get it based on it's configuration (what `Sensors` the `Thing` has on-board).

I have found the [work by Chih-Yuan Huang and Hsin-Hsien Chen](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6387111/) in figure 7, introducing a Gateway, with a lot of additional operations. I was hoping for something simpler (my Thing are low-powered and like to comms down to a minimum)

Right now, I have a registration application (Node app running in a Docker container, close to the FROST-Server, connected to the same database) that receives a registration message from the `Thing` each time it boots (can be send anytime).

```json
{
  "name": "DC3F7F",
  "type": "a",
  "sensors": [
    ["HMC5883", "Magnetic fields"], 
    ["BMP180","Barometric Pressure","Temperature"]]
}
```

`"name"`unique identifier of the Thing (last 3 bytes of the MAC)
`"type"` type information of the Thing
`"sensors"`array of Sensors and ObservedProperties (need to already exist in the Database). 1 element is Sensor name, then n elements for the ObservedProperties (n >= 1).

The registration application, upon receiving the Thing's startup message, checks if the Thing exists and, if needed, creates it (if the location is not given, it uses a default location). Then checks if the Sensors and ObservedProperties exist in the database (they must), and with the `Sensor` `ObservedProperties` combo creates a `Datastream` (if it does not exist). The `type` information is used to create specifics in the Thing's properties.

the registration service then returns:
```json
{
  "iot.id": 1234,
  "dateTime": "2022-03-25T13:25:00Z",
  "datastreams": [23, 64, 65]
}
```

`"iot.id"` is the iot.id of the (newly created) `Thing`
`"dateTime"` my Things have no RTC
`"datastreams"` iot.id's of the `Datastreams`, in the same order as the `Sensors` in the request (note that BMP180 has 2 `ObservedProperties` and thus 2 `Datastreams`) 

### My question
Are there any best-practices (or recommendations) on how to ([semi] automatically) register `Things` to get the` Datastream @iot.id?` when they announce themselves to the FROST-Server (either at startup or random moments).

When the amount of Things is limited, a simple admin tool (or even manually) can be used to create the Datastream - but when the number of Things grow, a more automated system is needed.

**Comment by hylkevds** [comment]
Currently there is no best practice for this.
In most of the use-cases that I work on the Thing does not directly communicate with the SensorThings API end-point, but with some service in-between. Either because the Thing has no direct network access (LoRa, Narrowband, Bluetooth, etc) or because we're pulling data from a completely separate service. In all of those cases the set of scripts that handle the translations also handle creating the entities in the SensorThings service.

In the end, this question comes down to: "what is the central point of truth where the sensors are registered". 
If that central point of truth is not the STA service, then whatever piece of software copies Sensor metadata from that main registry to the STA service has to create the correct entities.
If that central point of truth is the STA service, then the software (GUI) used to register sensors has to create the correct entities.

For things that do have network access, the first problem is finding the SensorThings API end-point in the first place.
Of course, once a Thing has found the SensorThings API end-point, it can "simply" do a query that returns the correct Datastreams. Something like:
```
v1.1/Datastreams?$filter=Thing/properties/mac eq 'DC:3F:7F:CD:F3:F1' and ObservedProperty/name eq 'Magnetic fields'
```
if need be, extended with the Sensor-ID:
```
v1.1/Datastreams?$filter=Thing/properties/mac eq 'DC:3F:7F:CD:F3:F1'
    and ObservedProperty/name eq 'Magnetic fields'
    and Sensor/name eq 'HMC5883'
```
